algviz/parsers/structures.py

```diff
@@ -0,0 +1,207 @@
+        # However, these __eq__ methods are very useful for testing
+        return isinstance(other, Pointer) and other.referent == self.referent
+
+    def untablify(self, obj_table):
+        self.referent = obj_table[self.referent]
+
+class Array(collections.UserList, DataStructure):
+    """An array data structure"""
+    def __init__(self, *args, **kwargs):
+        collections.UserList.__init__(self, *args)
+        DataStructure.__init__(self, **kwargs)
+
+    def untablify(self, obj_table):
+        # UserList has its underlying list accessible as a member
+        self.data = [obj_table[elt] for elt in self.data]
+
+class _Singleton(abc.ABCMeta):
+    # Embarassingly copied from https://stackoverflow.com/questions/6760685
+    _instances = {}
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            cls._instances[cls] = super(_Singleton, cls).__call__(*args, **kwargs)
+        return cls._instances[cls]
+
+class NullType(DataStructure, metaclass=_Singleton):
+    """A null value.  Similar to python's None.  Fun to make."""
+    def __init__(self):
+        super().__init__(uid="#null")
+        del self.metadata  # no static dictionary allowed
+
+    def untablify(self, obj_table):
+        pass
+
+    def __bool__(self):
+        return False
+
+    def __repr__(self):
+        return "Null"
+
+Null = NullType()
+
+class LinkedListNode(DataStructure):
+    def __init__(self, value, successor=Null):
+        self.value = value
+        self.successor = successor
+
+    def untablify(self, obj_table):
+        self.value = obj_table[self.value]
+        self.successor = obj_table[self.successor]
+
+class Graph(DataStructure):
+    def __init__(self, nodes, edges, **kwargs):
+        super().__init__(**kwargs)
+        self.nodes = nodes
+        self.edges = edges
+
+    def untablify(self, obj_table):
+        self.nodes = [obj_table[n] for n in self.nodes]
+        self.edges = [obj_table[e] for e in self.edges]
+
+class Node(DataStructure):
+    def __init__(self, data, **kwargs):
+        super().__init__(**kwargs)
+        self.data = data
+        # self.edges = edges  # no storing edges, I guess.  TODO: discuss this with people
+
+    def untablify(self, obj_table):
+        self.data = obj_table[self.data]
+        # self.edges = list(obj_table[e] for e in self.edges)
+
+class Edge(DataStructure):
+    def __init__(self, orig, dest, data=Null, **kwargs):
+        super().__init__(**kwargs)
+        self.orig = orig
+        self.dest = dest
+        self.data = data
+
+    def untablify(self, obj_table):
+        self.orig = obj_table[self.orig]
+        self.dest = obj_table[self.dest]
+        self.data = obj_table[self.data]
+
+class Widget(DataStructure):
+    # Potentially our most useful DataStructure
+
+    def __eq__(self, other):
+        return isinstance(other, Widget)
+
+    def untablify(self, obj_table):
+        pass
+
+class BinaryTree(DataStructure):
```

Comment:
Not immediately necessary, but non binary trees need to be added eventually

Reply:
Yeah.  I guess there's no need to have binary trees as a separate thing from n-ary trees.  For instance, a node with a right child but no left child could look like:
```json
{"type": "btnode", "data": 1, "children": ["#null", {"type": "btnode", "data": 2}]}
```

Reply:
Eventually we will need to separate the two cases because we visualize binary trees separately from generic trees. It would make the most sense to me for the user to define their tree as one or the other in case they want to visualize a binary tree as an arbitrary tree.

Reply:
Any drawing algorithm that works for arbitrary trees should work for binary trees as well.  So if binary trees are different at all, they should be a subclass.